### ISSUE TITLE
different metrics output

### ISSUE THREAD

**Issue by furycurry** [ultralytics/ultralytics#17646]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Greetings,

I want to get output of metrics P, R, AP50, AP75, mAP, mAP50, APs while validating my project. Unfortunately I cannot find a way to do so all at once. I have tried modifying the test.py like below

```
import torch
from ultralytics import YOLO

model = YOLO("yolov8n-EMA.yaml").load(
    '.../weights/best.pt'
) 

data_path = '.../data.yaml'

results = model.train(
    deterministic=False,
    data=data_path,
    optimizer='AdamW',
    weight_decay=1e-2,
    lr0=1e-4,
    batch=4,
    epochs=10,
)

val_results = model.val(data=data_path)

metrics = val_results.results_dict  # `results_dict` contains all the detection metrics
precision = metrics.get('P', None)       # Precision
recall = metrics.get('R', None)         # Recall
map50 = metrics.get('mAP50', None)      # Mean AP at IoU=0.5
map50_95 = metrics.get('mAP50-95', None)  # Mean AP at IoU=0.5:0.95
ap50 = metrics.get('AP50', None)        # AP at IoU=0.5
ap75 = metrics.get('AP75', None)        # AP at IoU=0.75
aps = metrics.get('APs', None)          # AP for small objects (if available)

print("Validation Metrics:")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"AP50: {ap50}")
print(f"AP75: {ap75}")
print(f"mAP50: {map50}")
print(f"mAP50-95: {map50_95}")
if aps is not None:
    print(f"APs: {aps}")

with open("metrics_output.txt", "w") as f:
    f.write(f"Precision: {precision}\n")
    f.write(f"Recall: {recall}\n")
    f.write(f"AP50: {ap50}\n")
    f.write(f"AP75: {ap75}\n")
    f.write(f"mAP50: {map50}\n")
    f.write(f"mAP50-95: {map50_95}\n")
    if aps is not None:
        f.write(f"APs: {aps}\n")
```

which is throwing an error. How can I achieve to do so?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @furycurry, thank you for reaching out to Ultralytics 🚀!

We appreciate your interest in obtaining detailed validation metrics for your project. Please note that this is an automated response, and an Ultralytics engineer will also assist you shortly.

In the meantime, we highly recommend visiting the [Ultralytics Docs](https://docs.ultralytics.com) to check for useful information and examples that might address your query. You can explore our [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage sections, which provide guidance on many common questions.

If you are reporting a potential 🐛 Bug with your code snippet, it would greatly help us if you could provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). This will enable us to better assist you in debugging the issue.

Should you have a ❓ Question related to custom training, please offer as much detail as possible. This includes dataset examples, training logs, or any other relevant information, and ensure you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our community discussions based on your preference: for real-time interaction, hop onto [Discord](https://ultralytics.com/discord) 🎧. Alternatively, for more in-depth dialogues, you might find our [Discourse](https://community.ultralytics.com) or [Subreddit](https://reddit.com/r/ultralytics) helpful.

## Upgrade

Ensure that your `ultralytics` package is up-to-date, as well as all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml). Verify this in a [**Python>=3.8**](https://www.python.org/) environment, with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO works smoothly in these verified environments. These come with all necessary dependencies, including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/):

- **Notebooks** with free GPU access: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. Follow the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. Follow the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. CI tests validate the operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu every 24 hours and upon every commit.

**Comment by Y-T-G** [comment]
You should use the correct key names

```python
In [4]: results.results_dict
Out[4]:
{'metrics/precision(B)': 0.6395486210744012,
 'metrics/recall(B)': 0.5367955375553028,
 'metrics/mAP50(B)': 0.6048726076857669,
 'metrics/mAP50-95(B)': 0.44582350082013184,
 'fitness': 0.46172841150669536}
```

AP75 and APs isn't calculated.

**Comment by furycurry** [comment]
hi @Y-T-G 

Thanks for the feedback. I am actually trying to get the output of AP, AP50, AP75, mAP, APs, P, R metrics of the model. Is there any way of doing so?

**Comment by Y-T-G** [comment]
You can run val with `save_json=True` and then use it with COCOAPI for evaluation. You will also have to convert your labels to COCO format.

**Comment by Y-T-G** [comment]
Yes. It's for evaluation of the metrics.

**Comment by furycurry** [comment]
Thanks @Y-T-G . This is my code now:
```
import torch
import numpy as np
from ultralytics import YOLO

# Load a pretrained model
model = YOLO("yolov8n-EMA.yaml").load('dataset/weights/best.pt')

# Define the dataset path
data_path = 'dataset/data.yaml'

# Train the model
results = model.train(
    deterministic=False,
    data=data_path,
    epochs=100,
)

# Validate the model to compute metrics
val_results = model.val(data=data_path)

# Extract metrics
metrics = {
    "Precision": val_results.box.map50,   # Precision at IoU=0.5
    "Recall": val_results.box.map,        # Recall at IoU=0.5:0.95
    "AP50": val_results.box.ap50,         # Average Precision at IoU=0.5
    "mAP50": val_results.box.map50,       # Mean Average Precision at IoU=0.5
    "mAP75": val_results.box.map75,         # Average Precision at IoU=0.75
    "mAP50-95": val_results.box.map,      # Mean Average Precision at IoU=0.5:0.95
    "APs": val_results.box.maps[0],       # AP for small objects
}

# Print the metrics
print("Validation Metrics:")
for metric, value in metrics.items():
    if isinstance(value, (list, np.ndarray)):
        # If the value is a list or array, iterate and print each element
        print(f"{metric}:")
        for i, v in enumerate(value):
            print(f"  Class {i}: {v:.4f}")
    else:
        # If the value is a scalar, print directly
        print(f"{metric}: {value:.4f}")

# Save metrics to a file
with open("metrics_output.txt", "w") as f:
    for metric, value in metrics.items():
        if isinstance(value, (list, np.ndarray)):
            f.write(f"{metric}:\n")
            for i, v in enumerate(value):
                f.write(f"  Class {i}: {v:.4f}\n")
        else:
            f.write(f"{metric}: {value:.4f}\n")

```
I am getting the validation output out of this but I think its not using COCOAPI. Where shall I make the change do you think to get the appropriate output?

**Comment by Y-T-G** [comment]
You have to use `save_json=True` like I said which will save the results to a JSON file. Then you have to use COCOAPI to evaluate it.

https://github.com/MiXaiLL76/faster_coco_eval

**Comment by furycurry** [comment]
Hello @Y-T-G. Will it be okay if I follow [this](https://github.com/ultralytics/ultralytics/issues/14222) method to get the APs (AP small) result?

**Comment by furycurry** [comment]
Thanks for the feedback @Y-T-G . I am very much new to this metrics and overly confused on how to use it. If you have any step by step guide to share for understanding how to use this in my model, it will be very much helpful.

**Comment by Y-T-G** [comment]
You can use `datumaro` to convert your YOLO annotations to COCO format. Or export it as COCO format in your labeling tool. Then run `model.val()` qith `save_json=True` which will save `predictions.json`. Then use this to evaluate it:

https://github.com/MiXaiLL76/faster_coco_eval

**Comment by furycurry** [comment]
Hello @Y-T-G. Kindly let me know if my understanding is wrong/right. So I need to train the model again with the converted dataset into coco format and run the `model.val()` with `save_json=True` which will be later used with **faster_coco_eval** to evaluate. Does it sound okay?

**Comment by Y-T-G** [comment]
No, you just need to run `model.val()`. You don't need to retrain.

**Comment by furycurry** [comment]
Thanks @Y-T-G . I was going through the [link](https://github.com/MiXaiLL76/faster_coco_eval) you suggested to use for evaluation. In one of the issues reported in that repo, it was mentioned that it doesn't print AP50, AP75, mAP, and APs. If you have worked with this evaluation metrics before, it will be very helpful if you can guide me.

**Comment by Y-T-G** [comment]
It does print it. It shows in the notebook.

https://github.com/MiXaiLL76/faster_coco_eval/blob/75fd297faf50baa892a6e09b6fa89b08831b7ab3/examples/comparison/ultralytics/colab_example.ipynb

`Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100` is APs

`Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ]` is AP75
